homeassistant/components/recorder/purge.py

```diff
@@ -1,88 +1,52 @@
 """Purge old data helper."""
-from datetime import timedelta
+from __future__ import annotations
+
+from datetime import datetime, timedelta
 import logging
 import time
+from typing import TYPE_CHECKING
 
 from sqlalchemy.exc import OperationalError, SQLAlchemyError
+from sqlalchemy.orm.session import Session
 
 import homeassistant.util.dt as dt_util
 
 from .models import Events, RecorderRuns, States
-from .util import execute, session_scope
+from .repack import repack_database
+from .util import session_scope
+
+if TYPE_CHECKING:
+    from . import Recorder
 
 _LOGGER = logging.getLogger(__name__)
 
+MAX_ROWS_TO_PURGE = 1000
+
 
-def purge_old_data(instance, purge_days: int, repack: bool) -> bool:
+def purge_old_data(instance: Recorder, purge_days: int, repack: bool) -> bool:
     """Purge events and states older than purge_days ago.
 
     Cleans up an timeframe of an hour, based on the oldest record.
     """
     purge_before = dt_util.utcnow() - timedelta(days=purge_days)
     _LOGGER.debug("Purging states and events before target %s", purge_before)
-
     try:
-        with session_scope(session=instance.get_session()) as session:
-            # Purge a max of 1 hour, based on the oldest states or events record
-            batch_purge_before = purge_before
-
-            query = session.query(States).order_by(States.last_updated.asc()).limit(1)
-            states = execute(query, to_native=True, validate_entity_ids=False)
-            if states:
-                batch_purge_before = min(
-                    batch_purge_before,
-                    states[0].last_updated + timedelta(hours=1),
-                )
-
-            query = session.query(Events).order_by(Events.time_fired.asc()).limit(1)
-            events = execute(query, to_native=True)
-            if events:
-                batch_purge_before = min(
-                    batch_purge_before,
-                    events[0].time_fired + timedelta(hours=1),
-                )
-
-            _LOGGER.debug("Purging states and events before %s", batch_purge_before)
-
-            deleted_rows = (
-                session.query(States)
-                .filter(States.last_updated < batch_purge_before)
-                .delete(synchronize_session=False)
-            )
-            _LOGGER.debug("Deleted %s states", deleted_rows)
-
-            deleted_rows = (
-                session.query(Events)
-                .filter(Events.time_fired < batch_purge_before)
-                .delete(synchronize_session=False)
-            )
-            _LOGGER.debug("Deleted %s events", deleted_rows)
-
-            # If states or events purging isn't processing the purge_before yet,
-            # return false, as we are not done yet.
-            if batch_purge_before != purge_before:
+        with session_scope(session=instance.get_session()) as session:  # type: ignore
```

Comment:
Although it won't remove the `type: ignore` comment, you could annotate `util.session_scope` with
```py
from homeassistant.helpers.typing import HomeAssistantType

@contextmanager
def session_scope(
    *, hass: Optional[HomeAssistantType] = None, session: Optional[Session] = None
) -> Generator[Session, None, None]:
```

I wouldn't annotate the whole `recorder.util` file just now.

Reply:
I think we are getting outside the scope of this fix so I'd prefer to leave it for now.